d2/src/main/java/com/linkedin/d2/balancer/dualread/UriPropertiesDualReadMonitor.java

```diff
@@ -0,0 +1,200 @@
+package com.linkedin.d2.balancer.dualread;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.linkedin.d2.balancer.properties.UriProperties;
+import com.linkedin.util.RateLimitedLogger;
+import com.linkedin.util.clock.SystemClock;
+import java.net.URI;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Function;
+import javax.annotation.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+public class UriPropertiesDualReadMonitor {
+  private static final Logger LOG = LoggerFactory.getLogger(UriPropertiesDualReadMonitor.class);
+
+  private final ConcurrentHashMap<String, ClusterMatchRecord> _clusters = new ConcurrentHashMap<>();
+  private final ConcurrentHashMap<String, Object> _clusterLocks = new ConcurrentHashMap<>();
```

Comment:
You don't need to do this, you can use the raw `ClusterMatchRecord` as the lock. It makes things simpler too since you can freely throw `synchronized` on every method of `ClusterMatchRecord` to ensure nothing weird happens. `toString`, `equals` and `hashCode` all access the fields, so they need to be locked too. Java's synchronized locks are reentrant, so this works no problem:
```java
ClusterMatchRecord cluster = _clusters.computeIfAbsent(clusterName, k -> new ClusterMatchRecord());
synchronized (cluster) {
  // In go this would fail since it attempts to acquire the lock again. In Java, this works
  System.out.println(cluster.toString());
}
```

Reply:
I hesitated about it since the cluster gets updated in the block (like cluster.uris = xxx), but that should be ok. Updated.